### PR TITLE
Create example to load `*.csv` and transfer to `HeteroData`

### PR DESCRIPTION
**Pull request by yaoyaowd**

### CODE REVIEW

File: examples/load_csv.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+
+class EmptyNodeData(object):
+    '''
+    The empty node data contains the node id with no features.
+    We use empty node data as a temporary storage to load edges
+    and generate ids from edges directly.
+    '''
+
+    def load_csv(self, name, filename, id_column=None):
```
Comment (chantat) [inline examples/load_csv.py:16]: Are you planning to add other data format here as well? If not, it probably makes more sense to consolidate this into a single function rather than a class because it is quite verbose:
`
empty_node_data = EmptyNodeData()
empty_node_data.load_csv(a, b, c, d)
node_storage = empty_node_data.to_node_storage()
`
Alternative:
`
node_storage = load_empty_node_data_from_csv(a, b, c, d)
`

File: examples/load_csv.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        self.name = ()
+        self.name = name
+        self.filename = filename
+        self.id_column = id_column
+
+        data = pd.read_csv(filename, low_memory=False)
+        self.invert_index = {
```
Comment (chantat) [inline examples/load_csv.py:53]: Name is a bit misleading. Maybe something like idx_to_id and id_to_idx would be better?

**Comment by rusty1s** [comment]
Thanks @yaoyaowd. Do you wanna keep that as an example or do you wanna create a tutorial in the documentation out of it?

**Comment by yaoyaowd** [comment]
@rusty1s @chantat It is ready to review now. Now I change the file to an ipython notebook include documentation.
https://github.com/vrtex-team/pytorch_geometric/blob/0420a3710831a9d671e04d0f78fd8e1a265eeb70/examples/data/load_csv.ipynb

**Comment by rusty1s** [comment]
Thanks @yaoyaowd. Instead of adding this as a Jupyter Notebook, what's your take on adding this to the documentation (similar to https://github.com/vrtex-team/pytorch_geometric/pull/83) and providing download options to run the script locally/in notebooks (similar to https://docs.dgl.ai/tutorials/models/1_gnn/1_gcn.html#sphx-glr-tutorials-models-1-gnn-1-gcn-py).

**Comment by rusty1s** [comment]
It's hard to give a review of a Notebook here, so I just list my comments here:

* We can probably automatically download the data via `torch_geometric.data.download_url`
* The tutorial should give a quick overview of the dataset and for what it is used for
* "I" should be replaced by "we"
* Is it possible to use the `transformers` library to encode the title and overview of the movies?
* I'm not sure why you need the `NodeData` and `EdgeData` objects as they do not have state. I suggest the interface to be the following: 
  * `load_movie_node_features` creates a node feature matrix `x` from a `pd.DataFrame`
  * `load_rating` creates a vector of link-level ground-truth labels
  * `load_edge_index` creates the `edge_index` connecting movies/users.
* After that, the `HeteroData` can be easily created via
```python
data = HeteroData()
data['movie].x = load_movie_node_features()
data['user'].num_nodes = ...
data['user', 'rates', movie].edge_index = load_edge_index()
data['user', 'rates', movie].rating = load_rating()
```
without of making explicit usage of `NodeStorage` and `EdgeStorage`, as in general, the user does not have to care about those classes in the first place.

**Comment by yaoyaowd** [comment]
I made the following change to the PR.

- The dataset changed to movie lens from group lens: https://grouplens.org/datasets/movielens/ Now the demo include download data from the url and it is more straightforward to run.

- Included a quick stats of the data and the link to group lens website.

- A few update on the document and code comments.

For the `class` vs `function`, I had a thought that in the future we can formalize the data loader class so that we can easily customize how users load csv file and transform the feature like following so that users do not need to worry about write their own loader.

```python
class NodeDataLoader(filename, id=...):
  def __init__(self):
    self.id = id
    pass # does not load the data

  def add_feature(feature_name, transfom_func):
    self.feature_transforms[feature_name] = transform_func

  def load():
    pass # load the data
```

File: docs/source/notes/loading_csv.rst [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+    from torch_geometric.data import download_url
+
+    DATA_URL = "https://files.grouplens.org/datasets/movielens/ml-latest-small.zip"
+    DATA_FOLDER = "."
+
+    data_path = download_url(DATA_URL, DATA_FOLDER)
+    print(f"MovieLens data downloaded to {data_dir}")
+
+    with zipfile.ZipFile(data_path, 'r') as zip_ref:
```
[the remark this answers is not in the document]
Comment (mrjel) [reply docs/source/notes/loading_csv.rst:25]: changed in doc

File: docs/source/notes/load_csv.rst [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      - Grumpier Old Men (1995)
+      - Comedy|Romance
+    * - 4
+      - Waiting to Exhale (1995)
+      - Comedy|Drama|Romance
+    * - 5
+      - Father of the Bride Part II (1995)
+      - Comedy
+
+We see that the :obj:`movies.csv` file provides three columns: :obj:`movieId` assigns a unique identifier to each movie, while the :obj:`title` and :obj:`genres` columns are represent the title and the genres of the given movie.
```
Comment (mrjel) [inline docs/source/notes/load_csv.rst:56]:
```suggestion
We see that the :obj:`movies.csv` file provides three columns: :obj:`movieId` assigns a unique identifier to each movie, while the :obj:`title` and :obj:`genres` columns represent title and genres of the given movie.
```

File: docs/source/notes/load_csv.rst [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+      - 964983815
+    * - 1
+      - 50
+      - 5.0
+      - 964982931
+
+The :obj:`ratings.csv` data connects users (as given by :obj:`userId`) and movies (as given by :obj:`movieId`), and defines how a given user has rated a specific movie (:obj:`rating`).
+Due to simplicity, we do not make use of the additional :obj:`timestamp` information.
+
+For representing this data in the PyTorch Geometric data format, we first define a method :meth:`load_node_csv` that reads in a :obj:`*.csv` file and returns a node-level feature representation :obj:`x` of shape :obj:`[num_nodes, num_features]` from it:
```
Comment (mrjel) [inline docs/source/notes/load_csv.rst:91]:
```suggestion
For representing this data in the PyTorch Geometric data format, we first define a method :meth:`load_node_csv` that reads in a :obj:`*.csv` file and returns a node-level feature representation :obj:`x` of shape :obj:`[num_nodes, num_features]`:
```

File: docs/source/notes/load_csv.rst [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            self.device = device
+            self.model = SentenceTransformer(model_name, device=device)
+
+        @torch.no_grad()
+        def __call__(self, df):
+            x = self.model.encode(df.values, show_progress_bar=True,
+                                  convert_to_tensor=True, device=self.device)
+            return x.cpu()
+
+The :class:`SequenceEncoder` class loads a pre-trained NLP model as given by :obj:`model_name`, and uses it to encode a list of strings into PyTorch tensor of shape :obj:`[num_strings, embedding_dim]`.
```
Comment (mrjel) [inline docs/source/notes/load_csv.rst:133]:
```suggestion
The :class:`SequenceEncoder` class loads a pre-trained NLP model as given by :obj:`model_name`, and uses it to encode a list of strings into a PyTorch tensor of shape :obj:`[num_strings, embedding_dim]`.
```

File: docs/source/notes/load_csv.rst [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+We can use this :class:`SequenceEncoder` to encde the :obj:`title` of the :obj:`movies.csv` file.
```
Comment (mrjel) [inline docs/source/notes/load_csv.rst:134]:
```suggestion
We can use this :class:`SequenceEncoder` to encode the :obj:`title` of the :obj:`movies.csv` file.
```

**Review by mrjel — APPROVED** [review]
